plugins/amdgpu/amdgpu_plugin_topology.c

```diff
@@ -703,3 +832,347 @@ int topology_parse(struct tp_system *sys, const char *message)
+				return false;
+			}
+		} else {
+			/* If GPUs have P2P-PCIe iolinks to this GPU, then at least one CPU will
+			 * also have a P2P-PCIe iolink to this GPU, so it seems that we do not need
+                         * to consider P2P-PCIe iolinks from GPU to GPU for now. Once P2P-PCIe
+                         * iolinks are exposed via p2p_links we may have to add additional code here
+                         * to validate P2P-PCIe links between GPUs. */
+		}
+	}
+	pr_debug("[0x%04X -> 0x%04X] Map is possible\n", src_node->gpu_id, dest_node->gpu_id);
+
+	if (!maps_add_gpu_entry(new_maps, src_node->gpu_id, dest_node->gpu_id)) {
+		/* This is a critical error because we are out of memory */
+		return false;
+	}
+	maps_print(new_maps);
+	return true;
+}
+
+static bool map_devices(struct tp_system *src_sys, struct tp_system *dest_sys, struct list_head *src_nodes,
+		struct list_head *dest_nodes, struct device_maps *maps)
+{
+	struct tp_node *src_node, *dest_node, *dest_node_tmp;
+	struct device_maps new_maps;
+
+	/* Pick the first src node from the list of nodes and look for a dest node that is mappable.
+	 * If we find a mappable destination node, then we add src node and dest node mapping to
+	 * device_maps and recursively try to map the remaining nodes in the list.
+	 * If there are no more src nodes in the list, then we have found a successful combination
+	 * of src to dest nodes that are mappable. */
+	if (list_empty(src_nodes)) {
+		pr_debug("All nodes mapped successfully\n");
+		return true;
+	}
+
+	pr_debug("Mapping list src nodes [%s]\n", mapping_list_to_str(src_nodes));
+	pr_debug("Mapping list dest nodes [%s]\n", mapping_list_to_str(dest_nodes));
+
+	src_node = list_first_entry(src_nodes, struct tp_node, listm_mapping);
+	pr_debug("Looking for match for node 0x%04X\n", src_node->gpu_id);
+
+	list_del(&src_node->listm_mapping);
+
+	list_for_each_entry_safe(dest_node, dest_node_tmp, dest_nodes, listm_mapping) {
+		maps_init(&new_maps);
+		if (map_device(src_sys, dest_sys, src_node, dest_node, maps, &new_maps)) {
+			pr_debug("Matched destination node 0x%04X\n", dest_node->gpu_id);
+
+			/* src node and dest node are mappable, add device_maps generated by
+			 * map_device to list of current valid device_maps, and recursively try to
+			 * map remaining nodes in the list. */
+
+			list_del(&dest_node->listm_mapping);
+			if (maps_append(maps, &new_maps))
+				return false;
+
+			if (map_devices(src_sys, dest_sys, src_nodes, dest_nodes, maps)) {
+				pr_debug("Matched nodes 0x%04X and after\n", dest_node->gpu_id);
+				return true;
+			} else {
+				/* We could not map remaining nodes in the list. Add dest node back
+				 * to list and try to map next dest ndoe in list to current src
+				 * node. */
+				pr_debug("Nodes after [0x%04X -> 0x%04X] did not match, adding list back\n",
+								src_node->gpu_id, dest_node->gpu_id);
+
+				list_add(&dest_node->listm_mapping, dest_nodes);
```

Comment:
This adds dest_node back at the end of the list. That means you're going to see it again later in this loop. I think you need to add it back in the original position. Otherwise it can break the recursion. E.g. imagine dest_nodes is A-B-C-D. Then you recurse:

Level 1: take node A off the list, list_for_each_entry_save stores pointer to next entry: dest_node_temp = B
  Level 2: take node B off the list, next is C
    ...
    add B back at the end of the list. dest_nodes is now D-C-B
  back in level 1, add A back to the end of the list, dest_nodes is now: D-C-B-A

next loop iteration uses B, which was stored in dest_node_tmp, which is now near the end of the list. D and C are skipped and never considered.

Reply:
Never mind. list_adds to the front, not the end. So I think it works OK.

Reply:
Yes. There is one of unit tests in a further commit that tests this situation.